File: tests/ui/presentation.test.ts
import { presentTool } from "../../src/ui/presentation/tools.js";
import { buildAgentRows, actionKey } from "../../src/ui/presentation/operations.js";
import type { WatcherRecord } from "../../src/schemas.js";
import type { TerminalPreview } from "../../src/ui/hooks/useTerminalPreview.js";

describe("presentTool", () => {
  it("maps known first-party tools to human verbs + targets", () => {
    expect(presentTool("fs.search", { query: "watcher" })).toEqual({
      label: "Searched",
      detail: "watcher",
    });
    expect(presentTool("agentTask.spawnForEdits", { title: "repair tests" })).toMatchObject({
      label: "Delegated",
      detail: "repair tests",
    });
    expect(presentTool("watcher.terminal.create", { goal: "wait for tests" })).toMatchObject({
      label: "Watching",
    });
  });

  it("falls back to the internal name (never raw fn syntax) for unknown tools", () => {
    const p = presentTool("some.exotic.tool", { a: 1 });
    expect(p.label).toBe("some.exotic.tool");
    expect(p.label).not.toContain("(");
  });
});

function watcher(over: Partial<WatcherRecord>): WatcherRecord {
  return {
    id: "wch_1",
    kind: "terminal",
    title: "repair tests",
    goal: "wait for tests",
    targetsJson: JSON.stringify(["term_8"]),
    cadenceMs: 1000,
    modelTier: "small",
    status: "active",
    lastClassification: "still_working",
    nextCheckAt: 0,
    createdAt: 0,
    ...over,
  };
}

describe("buildAgentRows", () => {
  it("merges a watcher with its terminal preview into one agent row", () => {
    const previews: TerminalPreview[] = [
      { terminalId: "term_8", tail: "line one\n  42 passed", updatedAt: 0, agentState: "working" },
    ];
    const rows = buildAgentRows([watcher({})], previews);
    expect(rows).toHaveLength(1);
    expect(rows[0].id).toBe("term_8");
    expect(rows[0].preview).toBe("42 passed");
    expect(rows[0].agentState).toBe("working");
  });

  it("orders rows by human urgency (needs-input before working)", () => {
    const rows = buildAgentRows([
      watcher({ id: "wch_working", lastClassification: "still_working" }),
      watcher({ id: "wch_input", lastClassification: "waiting_for_input" }),
    ]);
    expect(rows[0].watcherId).toBe("wch_input");
    expect(rows[0].needsAttention).toBe(true);
  });

  it("prefers the persisted lastEpistemicKind over the classification fallback (#85)", () => {
    // still_working would derive "inferred", but a stored kind is authoritative.
    const rows = buildAgentRows([
      watcher({ lastClassification: "still_working", lastEpistemicKind: "observed" }),
    ]);
    expect(rows[0].epistemicKind).toBe("observed");
  });

  it("falls back to a classification-derived kind for rows without lastEpistemicKind (#85)", () => {
    const [exited] = buildAgentRows([
      watcher({ id: "a", lastClassification: "terminal_exited" }),
    ]);
    const [working] = buildAgentRows([
      watcher({ id: "b", lastClassification: "still_working" }),
    ]);
    const [unknown] = buildAgentRows([watcher({ id: "c", lastClassification: "unknown" })]);
    expect(exited.epistemicKind).toBe("observed");
    expect(working.epistemicKind).toBe("inferred");
    expect(unknown.epistemicKind).toBe("unverified");
  });

  it("passes a raw (unvalidated) lastEpistemicKind straight through (#85)", () => {
    // rowToWatcher doesn't re-validate the stored string; a corrupt value degrades
    // safely downstream (epistemicMark returns null → no tag), so the row keeps it.
    const rows = buildAgentRows([
      watcher({ lastEpistemicKind: "bogus" as any }),
    ]);
    expect(rows[0].epistemicKind).toBe("bogus");
  });
});

describe("actionKey", () => {
  it("derives a key from the first letter of the label", () => {
    expect(actionKey("focus terminal", 0)).toBe("F");
    expect(actionKey("rerun", 1)).toBe("R");
  });
});

File: internal/ui/presentation_test.go
package ui

import (
	"testing"

	"github.com/daintreehq/assistant/internal/domain"
)

// presentation_test.go: presentTool maps first-party tools to human verbs (never
// raw fn() syntax for unknowns), and
// BuildAgentRows merges/orders watchers + threads the epistemic kind (persisted
// preference, classification fallback, raw passthrough).

func TestPresentTool_FirstPartyVerbs(t *testing.T) {
	cases := map[string]string{
		"fs.search":               "Searched",
		"agentTask.spawnForEdits": "Delegated",
		"watcher.terminal.create": "Watching",
		"fs.read":                 "Read",
		"fs.list":                 "Listed",
		// The scheduling verb is keyed on the real tool name.
		"timer.schedule": "Scheduled",
		// High-frequency orchestration tools that used to fall through to their raw
		// dotted name in the activity tree (the scratch store + await/send family).
		"terminal.awaitAll":    "Waited",
		"terminal.sendCommand": "Ran",
		"scratch.create":       "Scratch",
		"scratch.set":          "Scratch",
		"scratch.get":          "Scratch",
		"memory.recall":        "Recalled",
		"artifact.read":        "Read artifact",
	}
	for name, want := range cases {
		if got := presentTool(name); got != want {
			t.Errorf("presentTool(%q) = %q, want %q", name, got, want)
		}
	}
}

// TestPresentTool_NoRawDottedNames guards against the presentation gap that motivated
// the verb-map expansion: a first-party tool that renders its raw dotted name (e.g.
// "scratch.create") next to nicely-verbed neighbours ("Delegated"). Every canonical
// local tool name should resolve to a human label — never a string still carrying a
// dot from its namespace.
func TestPresentTool_NoRawDottedNames(t *testing.T) {
	names := []string{
		"scratch.create", "scratch.set", "scratch.get", "scratch.delete", "scratch.drop",
		"terminal.awaitAll", "terminal.sendCommand", "terminal.rename", "terminal.close",
		"terminal.arm", "terminal.disarm", "terminal.disarmAll", "terminal.moveToWorktree",
		"terminal.run.async", "terminal.await.async", "async.list", "async.cancel",
		"memory.recall", "memory.list", "memory.save", "memory.forget", "memory.pin", "memory.unpin",
		"artifact.read", "copyTree.generate", "copyTree.injectToTerminal",
		"forge.getPR", "watcher.watchPR", "audit.export",
		"agentTask.status", "agentTask.list", "agentTask.superviseTerminal",
		"user.askMultipleChoice",
		"workflow.plan", "workflow.reconcile", "workflow.create", "workflow.list",
	}
	for _, n := range names {
		got := presentTool(n)
		if got == n || contains(got, ".") {
			t.Errorf("presentTool(%q) = %q — still a raw dotted name; add a human verb", n, got)
		}
	}
}

func TestPresentTool_UnknownFallsBackToInternalName(t *testing.T) {
	// Unknown tools fall back to the RAW internal name — NEVER raw "fn(" syntax, and
	// NOT title-cased.
	got := presentTool("some.exotic.tool")
	if got != "some.exotic.tool" {
		t.Errorf("presentTool unknown leaf = %q, want raw internal name some.exotic.tool", got)
	}
	for _, name := range []string{"some.exotic.thing", "weird"} {
		if p := presentTool(name); contains(p, "(") {
			t.Errorf("presentTool(%q) = %q must not contain a paren (raw fn syntax)", name, p)
		}
	}
}

func contains(s, sub string) bool {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}

// watcherRec builds a watcher record with overridable classification/epistemic.
func watcherRec(id string, classification string, epistemic *domain.EpistemicKind) domain.WatcherRecord {
	w := domain.WatcherRecord{
		ID:        id,
		Kind:      "terminal",
		Title:     "repair tests",
		Goal:      "wait for tests",
		Status:    "active",
		CadenceMs: 1000,
	}
	if classification != "" {
		c := classification
		w.LastClassification = &c
	}
	w.LastEpistemicKind = epistemic
	return w
}

func TestBuildAgentRows_OrdersByUrgency(t *testing.T) {
	rows := BuildAgentRows([]domain.WatcherRecord{
		watcherRec("wch_working", string(domain.ClassStillWorking), nil),
		watcherRec("wch_input", string(domain.ClassWaitingForInput), nil),
	}, nil, nil)
	if len(rows) != 2 {
		t.Fatalf("rows = %d, want 2", len(rows))
	}
	// Needs-input sorts ahead of still-working.
	if rows[0].ID != "wch_input" {
		t.Errorf("rows[0] = %q, want wch_input (needs-input first)", rows[0].ID)
	}
	if !rows[0].NeedsAttention {
		t.Error("needs-input row should flag NeedsAttention")
	}
}

func TestBuildAgentRows_PrefersPersistedEpistemicKind(t *testing.T) {
	// still_working derives "observed" (deterministic working bypass), but a stored
	// kind is authoritative — here a persisted "inferred" overrides the derivation.
	inf := domain.EpistemicInferred
	rows := BuildAgentRows([]domain.WatcherRecord{
		watcherRec("a", string(domain.ClassStillWorking), &inf),
	}, nil, nil)
	if rows[0].EpistemicKind != domain.EpistemicInferred {
		t.Errorf("epistemicKind = %q, want inferred (persisted wins)", rows[0].EpistemicKind)
	}
}

func TestBuildAgentRows_ClassificationFallback(t *testing.T) {
	cases := []struct {
		classification string
		want           domain.EpistemicKind
	}{
		{string(domain.ClassTerminalExited), domain.EpistemicObserved},
		{string(domain.ClassStillWorking), domain.EpistemicObserved},
		{string(domain.ClassUnknown), domain.EpistemicUnverified},
	}
	for _, c := range cases {
		rows := BuildAgentRows([]domain.WatcherRecord{watcherRec("x", c.classification, nil)}, nil, nil)
		if rows[0].EpistemicKind != c.want {
			t.Errorf("classification %q → %q, want %q", c.classification, rows[0].EpistemicKind, c.want)
		}
	}
}

func TestBuildAgentRows_RawEpistemicPassthrough(t *testing.T) {
	// A corrupt stored kind degrades safely (epistemicTag returns "" → no tag) but is
	// not re-validated away — the row keeps it verbatim.
	bogus := domain.EpistemicKind("bogus")
	rows := BuildAgentRows([]domain.WatcherRecord{watcherRec("a", string(domain.ClassStillWorking), &bogus)}, nil, nil)
	if rows[0].EpistemicKind != "bogus" {
		t.Errorf("raw epistemicKind = %q, want passthrough bogus", rows[0].EpistemicKind)
	}
}
